### ISSUE TITLE
dev experience when passing around ResourceOptions is suboptimal

### ISSUE THREAD

**Issue by DmitryEfimenko** [angular/angular#58941]
### Description

Writing APIs that accept `ResourceOptions` as an argument is suboptimal. Not sure if there could anything be done about it or if it's just a limitation of TypeScript. See example below:

When using the following signature, all types are figured out correctly:
```ts
const resourceRef = resource({
  request: () => {
    const id = numberSignal();
    if (id == null) {
      return undefined;
    }
    return { id };
  },
  loader: ({ request: { id } }) => {
    return Promise.resolve(String(id))
  }
})
```
The type of `id` in the loader's request argument is correctly inferred as `number`. The type of `resourceRef` is `ResourceRef<string>`.
Notice how I didn't have to specify any types explicitly. Everything just works.

However, if options are declared separately, things fall apart:

```ts
const resourceOptions = {
  request: () => {
    const id = numberSignal();
    if (id == null) {
      return undefined;
    }
    return { id };
  },
  loader: ({ request: { id } }) => {
    return Promise.resolve(String(id))
  }
} satisfies ResourceOptions<any, any>;

const resourceRef = resource(resourceOptions);
```

With this, the type of resourceRef is correct: `ResourceRef<string>`. However, the type of `id` in loader's request argument is `any`.

Also, this might have something to do with the comment about an issue with `NoInfer`: https://github.com/angular/angular/issues/58871#issuecomment-2498195267

### Proposed solution

Making generic types optional by  providing default `any` value makes it slightly easier to work with, but overall does not provide the dev experience I'm after.

```ts
export interface ResourceOptions<T = any, R = any> {
  request?: () => R;
  loader: ResourceLoader<T, R>;
  equal?: ValueEqualityFn<T>;
  injector?: Injector;
}
```

Need guidance from someone more experienced with TypeScript.

### Alternatives considered

Create a function that takes options and returns them. This seems silly since it seemingly just adds boilerplate, but it works:

```ts
function createResourceOptions<T, R>(options: ResourceOptions<T, R>): ResourceOptions<T, R> {
  return options;
}

const resourceOptions = createResourceOptions({
  request: () => {
    const id = numberSignal();
    if (id == null) {
      return undefined;
    }
    return { id };
  },
  loader: ({ request: { id } }) => {
    return Promise.resolve(String(id))
  }
});

const resourceRef = resource(resourceOptions);
```
All types are figured out correctly.

**Comment by DmitryEfimenko** [comment]
right, providing types for ResourceOptions will make it work. But that defeats the whole thing that I'm trying to achieve, which is to avoid boilerplate of providing these generic types and let the types just be automatically inferred.

The reason I had `satisfies` in there is to give TS some guidance about what kind of object that should be - the `ResourceOptions` kind without providing all the other type information, which TS should be able to figure out based on the contents of the `request` and `loader` functions.

Maybe that's impossible today with TS... I don't know.

**Comment by alxhub** [comment]
tl;dr these types are designed to provide good type inference when used inline, and aren't meant to be extracted as separate variables.

In general, Angular doesn't expect or design for the "options" object types to be passed around independently of calls to their corresponding methods. Often those types are just declared inline, and if they're extracted as an interface, it's usually so they can be better visualized or navigated in the API docs.

This is just the nature of TS types for the signatures of such functions - there's nothing really we can do to remedy the situation. TypeScript themselves have reinforced that default type arguments and type inference do not mix - in situations where TypeScript relies on the default arguments, it won't be doing inference.
